File: app/src/main/java/info/horriblesubs/sher/ui/schedule/Schedule.java
package info.horriblesubs.sher.ui.schedule;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.os.Handler;
import android.view.View;
import android.widget.FrameLayout;

import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdSize;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.InterstitialAd;
import com.google.android.material.tabs.TabLayout;

import java.util.Calendar;
import java.util.List;
import java.util.Random;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.ViewPager;
import info.horriblesubs.sher.R;
import info.horriblesubs.sher.common.Change;
import info.horriblesubs.sher.common.DataTask;
import info.horriblesubs.sher.common.Navigate;
import info.horriblesubs.sher.model.base.LatestItem;
import info.horriblesubs.sher.model.base.PageItem;
import info.horriblesubs.sher.model.base.ScheduleItem;
import info.horriblesubs.sher.model.response.ScheduleResponse;
import info.horriblesubs.sher.model.response.ShowsResponse;
import info.horriblesubs.sher.ui.schedule.fragment.Day;
import info.horriblesubs.sher.util.DialogX;

import static info.horriblesubs.sher.AppController.instance;

@SuppressLint("StaticFieldLeak")
public class Schedule extends AppCompatActivity
        implements DataTask.OnDataUpdate, Navigate.Task, Change {

    private InterstitialAd interstitialAd;
    private ViewPager viewPager;
    private TabLayout tabLayout;
    private DataTask dataTask;
    private Navigate navigate;
    private DialogX dialogX;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_b);
        viewPager = findViewById(R.id.viewPager);
        tabLayout = findViewById(R.id.tabLayout);
        tabLayout.setTabMode(TabLayout.MODE_SCROLLABLE);
        dataTask = new DataTask(this);
        navigate = new Navigate(this);
        onAdShow();
        for (String s : getResources().getStringArray(R.array.days))
            tabLayout.addTab(tabLayout.newTab().setText(s));
        viewPager.addOnPageChangeListener(new TabLayout.TabLayoutOnPageChangeListener(tabLayout));
        tabLayout.addOnTabSelectedListener(new TabLayout.ViewPagerOnTabSelectedListener(viewPager));
        onStartActivity();
    }

    @Override
    protected void onResume() {
        super.onResume();
        navigate.onResumeActivity();
    }

    private void onStartActivity() {
        List<ScheduleItem> list = instance.data.scheduleItems;
        if (list == null || list.size() == 0) {
            fetchData();
            return;
        }
        onScheduleUpdated(false, new ScheduleResponse(list));
    }

    private void onAdShow() {
        FrameLayout layout = findViewById(R.id.adViewLayout);
        AdRequest adRequest = new AdRequest.Builder().build();
        AdView adView = new AdView(this);
        adView.setAdUnitId(getString(R.string.ad_unit_footer_3));
        adView.setAdSize(AdSize.SMART_BANNER);
        layout.addView(adView);
        adView.loadAd(adRequest);

        int i = new Random().nextInt(100);
        interstitialAd = new InterstitialAd(this);
        if (i < 34)
            interstitialAd.setAdUnitId(getString(R.string.ad_unit_interstitial_1));
        else if (i < 68)
            interstitialAd.setAdUnitId(getString(R.string.ad_unit_interstitial_2));
        else if (i < 101)
            interstitialAd.setAdUnitId(getString(R.string.ad_unit_interstitial_3));
        interstitialAd.loadAd(new AdRequest.Builder().build());
        interstitialAd.setAdListener(new AdListener() {
            @Override
            public void onAdFailedToLoad(int e) {
                interstitialAd.loadAd(new AdRequest.Builder().build());
            }

            @Override
            public void onAdLoaded() {
                new Handler().postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        interstitialAd.show();
                    }
                }, 650);
            }
        });
    }

    @Override
    public void fetchData() {
        dataTask.fetchSchedule();
    }

    @Override
    public void preDataUpdate() {
        dialogX = new DialogX(this, true);
        dialogX.show();
    }

    @Override
    public void onLatestUpdated(boolean b, List<LatestItem> items) {

    }

    @Override
    public void onFavouritesUpdated(boolean b, List<PageItem> items) {

    }

    @Override
    public void onScheduleUpdated(boolean b, ScheduleResponse response) {
        if (!b && response != null && response.schedule != null && response.schedule.size() > 0) {
            instance.data.scheduleItems = response.schedule;
            viewPager.setAdapter(new PagerAdapter(getSupportFragmentManager(), response));
            viewPager.setCurrentItem(Calendar.getInstance().get(Calendar.DAY_OF_WEEK) - 1);
        } else
            navigate.setError(View.VISIBLE);
        if (dialogX != null)
            dialogX.dismiss();
    }

    @Override
    public void onShowsUpdated(boolean b, ShowsResponse showsResponse) {

    }

    @Override
    public void onThemeChange(boolean b) {
        if (b) {
            tabLayout.setTabTextColors(getResources().getColorStateList(R.color.tab_yellow));
            tabLayout.setSelectedTabIndicatorColor(getResources().getColor(R.color.yellow));
            tabLayout.setTabRippleColorResource(R.color.yellowRipple);
            tabLayout.setBackgroundResource(R.color.primaryDark);
        } else {
            tabLayout.setTabTextColors(getResources().getColorStateList(R.color.tab_blue));
            tabLayout.setSelectedTabIndicatorColor(getResources().getColor(R.color.blue));
            tabLayout.setTabRippleColorResource(R.color.blueRipple);
            tabLayout.setBackgroundResource(android.R.color.white);
        }
        if (viewPager.getAdapter() != null)
            viewPager.getAdapter().notifyDataSetChanged();
    }

    @Override
    public int getActivity() {
        return 2;
    }

    class PagerAdapter extends FragmentPagerAdapter {

        private final ScheduleResponse scheduleResponse;

        PagerAdapter(FragmentManager fragmentManager, ScheduleResponse scheduleResponse) {
            super(fragmentManager);
            this.scheduleResponse = scheduleResponse;
        }

        @Override
        public int getItemPosition(@NonNull Object object) {
            return POSITION_NONE;
        }

        @Override
        public Fragment getItem(int position) {
            return Day.newInstance(scheduleResponse, position);
        }

        @Override
        public int getCount() {
            return 8;
        }
    }
}

File: app/src/main/java/info/horriblesubs/sher/ui/main/schedule/Schedule.kt
package info.horriblesubs.sher.ui.main.schedule

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.PopupMenu
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.RecyclerView.Adapter
import androidx.viewpager2.widget.ViewPager2
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator
import info.horriblesubs.sher.R
import info.horriblesubs.sher.adapter.ItemClick
import info.horriblesubs.sher.api.horrible.model.ItemSchedule
import info.horriblesubs.sher.common.Info
import info.horriblesubs.sher.common.LoadingListener
import info.horriblesubs.sher.common.MenuHandler
import info.horriblesubs.sher.dialog.InfoDialog
import info.horriblesubs.sher.dialog.LoadingDialog
import info.horriblesubs.sher.ui.show.Show
import java.util.*
import kotlin.collections.ArrayList

class Schedule: Fragment(), LoadingListener, Observer<Map<String, List<ItemSchedule>>?>,
    PopupMenu.OnMenuItemClickListener, ItemClick<ItemSchedule> {

    private val dayAdapter: DaysAdapter = DaysAdapter(this)
    private var dLoading: LoadingDialog? = null
    private var handler: MenuHandler? = null
    private var vm: ScheduleVM? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        vm = ViewModelProvider(activity as AppCompatActivity).get(ScheduleVM::class.java)
        vm?.initialize(this)
    }

    override fun onCreateView(inflater: LayoutInflater, group: ViewGroup?, bundle: Bundle?): View? {
        return inflater.inflate(R.layout._b_fragment_3, group, false)
    }

    override fun onViewCreated(view: View, bundle: Bundle?) {
        super.onViewCreated(view, bundle)
        val viewPager: ViewPager2 = view.findViewById(R.id.viewPager)
        val tabLayout: TabLayout = view.findViewById(R.id.tabLayout)
        viewPager.adapter = dayAdapter
        TabLayoutMediator(tabLayout, viewPager, TabLayoutMediator.TabConfigurationStrategy{t, p ->
            t.text = dayAdapter.days[p].day
        }).attach()
        handler = MenuHandler(this, view, R.menu.menu_c)
        viewPager.orientation = ViewPager2.ORIENTATION_HORIZONTAL
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        vm?.schedule?.observe(viewLifecycleOwner, this)
        vm?.refresh()
    }

    override fun onChanged(result: Map<String, List<ItemSchedule>>?) {
        dayAdapter.result = result
        stop()
    }

    override fun onDestroy() {
        super.onDestroy()
        vm?.stop()
        stop()
    }

    override fun start() {
        if(dLoading == null)
            dLoading = context?.let{LoadingDialog(it)}
        dLoading?.show()
    }

    override fun stop() {
        dLoading?.dismiss()
        dLoading = null
    }

    override fun onMenuItemClick(item: MenuItem?): Boolean {
        val info: InfoDialog? = context?.let{InfoDialog(it)}
        when(item?.itemId) {
            R.id.info -> info?.show(Info.SCHEDULE, vm?.result?.value)
            R.id.refresh -> vm?.refresh(true)
            else -> return false
        }
        return true
    }

    override fun onClick(e: ItemSchedule?) {
        if (e == null) return
        val intent = Intent(activity, Show::class.java)
        intent.putExtra("link", e.link)
        startActivity(intent)
    }
}

private class DaysAdapter (private val itemClick: ItemClick<ItemSchedule>): Adapter<Day>() {
    var result: Map<String, List<ItemSchedule>>? = null
        set(value) {
            days = days()
            field = value
            notifyDataSetChanged()
        }

    var days: MutableList<Days> = this.days()
        private set

    private fun days(): MutableList<Days> {
        val i = Calendar.getInstance().get(Calendar.DAY_OF_WEEK)
        val list:MutableList<Days> = ArrayList()
        for (j in i..7)
            list.add(Days.from(j))
        for (j in 1 until i)
            list.add(Days.from(j))
        list.add(Days.OTHERS)
        return list
    }

    override fun onCreateViewHolder(group: ViewGroup, i: Int): Day {
        return Day(group)
    }

    override fun onBindViewHolder(holder: Day, position: Int) {
        holder.load(result?.get(days[position].day), itemClick)
    }

    override fun getItemCount(): Int {
        return days.size
    }
}
